File: packages/sdk/rust/packages/admin/src/vectorize.rs
//! VectorizeClient — Vectorize index access for server-side use.
//! Note: Vectorize is Edge-only. In local/Docker, the server returns stub responses.

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;

use crate::generated::admin_api_core::GeneratedAdminApi;
use edgebase_core::error::Error;
use edgebase_core::http_client::HttpClient;

/// Client for a user-defined Vectorize index.
pub struct VectorizeClient {
    http: Arc<HttpClient>,
    pub(crate) index: String,
}

impl VectorizeClient {
    pub(crate) fn new(http: Arc<HttpClient>, index: &str) -> Self {
        Self {
            http,
            index: index.to_string(),
        }
    }

    fn core(&self) -> GeneratedAdminApi<'_> {
        GeneratedAdminApi::new(&self.http)
    }

    /// Insert or update vectors.
    /// Returns mutation result with `ok`, optional `count` and `mutationId`.
    pub async fn upsert(&self, vectors: &[Value]) -> Result<Value, Error> {
        let body = serde_json::json!({"action": "upsert", "vectors": vectors});
        self.core().vectorize_operation(&self.index, &body).await
    }

    /// Insert vectors (errors on duplicate ID — server returns 409).
    /// Returns mutation result with `ok`, optional `count` and `mutationId`.
    pub async fn insert(&self, vectors: &[Value]) -> Result<Value, Error> {
        let body = serde_json::json!({"action": "insert", "vectors": vectors});
        self.core().vectorize_operation(&self.index, &body).await
    }

    /// Search for similar vectors.
    pub async fn search(
        &self,
        vector: &[f64],
        top_k: usize,
        filter: Option<&Value>,
        namespace: Option<&str>,
        return_values: Option<bool>,
        return_metadata: Option<&str>,
    ) -> Result<Vec<HashMap<String, Value>>, Error> {
        let mut body = serde_json::json!({
            "action": "search",
            "vector": vector,
            "topK": top_k,
        });
        if let Some(f) = filter {
            body["filter"] = f.clone();
        }
        if let Some(ns) = namespace {
            body["namespace"] = serde_json::json!(ns);
        }
        if let Some(rv) = return_values {
            body["returnValues"] = serde_json::json!(rv);
        }
        if let Some(rm) = return_metadata {
            body["returnMetadata"] = serde_json::json!(rm);
        }
        let res: Value = self.core().vectorize_operation(&self.index, &body).await?;
        extract_matches(&res)
    }

    /// Search by an existing vector's ID (Vectorize v2 only).
    pub async fn query_by_id(
        &self,
        vector_id: &str,
        top_k: usize,
        filter: Option<&Value>,
        namespace: Option<&str>,
        return_values: Option<bool>,
        return_metadata: Option<&str>,
    ) -> Result<Vec<HashMap<String, Value>>, Error> {
        let mut body = serde_json::json!({
            "action": "queryById",
            "vectorId": vector_id,
            "topK": top_k,
        });
        if let Some(f) = filter {
            body["filter"] = f.clone();
        }
        if let Some(ns) = namespace {
            body["namespace"] = serde_json::json!(ns);
        }
        if let Some(rv) = return_values {
            body["returnValues"] = serde_json::json!(rv);
        }
        if let Some(rm) = return_metadata {
            body["returnMetadata"] = serde_json::json!(rm);
        }
        let res: Value = self.core().vectorize_operation(&self.index, &body).await?;
        extract_matches(&res)
    }

    /// Retrieve vectors by their IDs.
    pub async fn get_by_ids(&self, ids: &[&str]) -> Result<Vec<HashMap<String, Value>>, Error> {
        let body = serde_json::json!({"action": "getByIds", "ids": ids});
        let res: Value = self.core().vectorize_operation(&self.index, &body).await?;
        if let Some(vectors) = res.get("vectors").and_then(|v| v.as_array()) {
            Ok(vectors
                .iter()
                .filter_map(|v| {
                    v.as_object()
                        .map(|o| o.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
                })
                .collect())
        } else {
            Ok(vec![])
        }
    }

    /// Delete vectors by IDs.
    /// Returns mutation result with `ok`, optional `count` and `mutationId`.
    pub async fn delete(&self, ids: &[&str]) -> Result<Value, Error> {
        let body = serde_json::json!({"action": "delete", "ids": ids});
        self.core().vectorize_operation(&self.index, &body).await
    }

    /// Get index info (vector count, dimensions, metric).
    pub async fn describe(&self) -> Result<Value, Error> {
        let body = serde_json::json!({"action": "describe"});
        self.core().vectorize_operation(&self.index, &body).await
    }
}

fn extract_matches(res: &Value) -> Result<Vec<HashMap<String, Value>>, Error> {
    if let Some(matches) = res.get("matches").and_then(|v| v.as_array()) {
        Ok(matches
            .iter()
            .filter_map(|v| {
                v.as_object()
                    .map(|o| o.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
            })
            .collect())
    } else {
        Ok(vec![])
    }
}

File: packages/sdk/js/packages/admin/src/vectorize.ts
/**
 * VectorizeClient — SDK client for user-defined Vectorize indexes
 *
 * Usage:
 *   const admin = createAdminClient(url, { serviceKey });
 *   await admin.vector('embeddings').upsert([{ id: 'doc-1', values: [...], metadata: { title: 'Hello' } }]);
 *   const results = await admin.vector('embeddings').search([0.1, 0.2, ...], { topK: 10 });
 *
 * Note: Vectorize is Cloudflare Edge-only. In local/Docker, the server returns stub responses.
 */
import type { HttpClient } from '@edge-base/core';
import { HttpClientAdapter } from '@edge-base/core';
import { DefaultAdminApi } from './generated/admin-api-core.js';

export interface VectorSearchOptions {
  topK?: number;
  filter?: Record<string, unknown>;
  namespace?: string;
  returnValues?: boolean;
  returnMetadata?: boolean | 'all' | 'indexed' | 'none';
}

export interface VectorSearchResult {
  id: string;
  score: number;
  values?: number[];
  metadata?: Record<string, unknown>;
  namespace?: string;
}

export interface VectorResult {
  id: string;
  values?: number[];
  metadata?: Record<string, unknown>;
  namespace?: string;
}

export interface VectorMutationResult {
  ok: true;
  count?: number;
  mutationId?: string;
}

export interface VectorIndexInfo {
  vectorCount: number;
  dimensions: number;
  metric: string;
  id?: string;
  name?: string;
  processedUpToDatetime?: string;
  processedUpToMutation?: string;
}

export class VectorizeClient {
  private adminCore: DefaultAdminApi;
  private index: string;

  constructor(httpClient: HttpClient, index: string) {
    this.adminCore = new DefaultAdminApi(new HttpClientAdapter(httpClient));
    this.index = index;
  }

  /** Insert or update vectors. */
  async upsert(vectors: Array<{ id: string; values: number[]; metadata?: Record<string, unknown>; namespace?: string }>): Promise<VectorMutationResult> {
    const res = await this.adminCore.vectorizeOperation(this.index, { action: 'upsert', vectors });
    return res as VectorMutationResult;
  }

  /** Insert vectors (errors on duplicate ID). */
  async insert(vectors: Array<{ id: string; values: number[]; metadata?: Record<string, unknown>; namespace?: string }>): Promise<VectorMutationResult> {
    const res = await this.adminCore.vectorizeOperation(this.index, { action: 'insert', vectors });
    return res as VectorMutationResult;
  }

  /** Search for similar vectors. */
  async search(
    vector: number[],
    options?: VectorSearchOptions,
  ): Promise<VectorSearchResult[]> {
    const body: Record<string, unknown> = {
      action: 'search',
      vector,
      topK: options?.topK,
      filter: options?.filter,
    };
    if (options?.namespace !== undefined) body.namespace = options.namespace;
    if (options?.returnValues !== undefined) body.returnValues = options.returnValues;
    if (options?.returnMetadata !== undefined) body.returnMetadata = options.returnMetadata;

    const res = await this.adminCore.vectorizeOperation(this.index, body) as { matches: VectorSearchResult[] };
    return res.matches;
  }

  /** Search by an existing vector's ID (Vectorize v2 only). */
  async queryById(
    vectorId: string,
    options?: VectorSearchOptions,
  ): Promise<VectorSearchResult[]> {
    const body: Record<string, unknown> = {
      action: 'queryById',
      vectorId,
      topK: options?.topK,
      filter: options?.filter,
    };
    if (options?.namespace !== undefined) body.namespace = options.namespace;
    if (options?.returnValues !== undefined) body.returnValues = options.returnValues;
    if (options?.returnMetadata !== undefined) body.returnMetadata = options.returnMetadata;

    const res = await this.adminCore.vectorizeOperation(this.index, body) as { matches: VectorSearchResult[] };
    return res.matches;
  }

  /** Retrieve vectors by their IDs. */
  async getByIds(ids: string[]): Promise<VectorResult[]> {
    const res = await this.adminCore.vectorizeOperation(this.index, { action: 'getByIds', ids }) as { vectors: VectorResult[] };
    return res.vectors;
  }

  /** Delete vectors by IDs. */
  async delete(ids: string[]): Promise<VectorMutationResult> {
    const res = await this.adminCore.vectorizeOperation(this.index, { action: 'delete', ids });
    return res as VectorMutationResult;
  }

  /** Get index info (vector count, dimensions, metric). */
  async describe(): Promise<VectorIndexInfo> {
    const res = await this.adminCore.vectorizeOperation(this.index, { action: 'describe' });
    return res as VectorIndexInfo;
  }
}
